### PR TITLE
Packages: Add common utilities to WP.com block editor package

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

Add a new folder for all the common utilities we need to run in both block editor environments (Gutenframe and WP Admin). 

So far, these utilities are the rich text extensions to add the missing formats in Core (justify, underline) and the "Switch to Classic" button on the "More tools" menu.

The logic of these utilities were previously maintained on the `gutenberg-wpcom.js` file, so the files added to the `common/` folder on this PR contain the same code but using ES2015+ syntax.

The documentation has also changed in order to make more obvious there are 2 environments we need to support. Note that I also grouped the previous scripts in a `calypso/` folder to reflect that the utilities inside this folder will be run only in a iframed block editor context.

Given that this is adding a new entry file to the build script, I decided to split the `bundle` script into 3 smaller `build` scripts (one for each entry file), so we don't have a very long script command. The `build` scripts runs now all the smaller build scripts in parallel (using `npm-run-all`) and requires an explicit output path (reason why the `clean` script is not needed anymore).

#### Testing instructions

Check D26970-code

### CODE REVIEW

File: apps/wpcom-block-editor/README.md [hunk 1]
```diff
@@ -4,36 +4,37 @@ This package provides utilities for the WordPress.com block editor integration.
...
-### `iframe-bridge-server.js`
+- **WP Admin block editor**. The block editor loaded by the WP Admin interface on `https://<SITE_SLUG>/wp-admin/post-new.php`.
+- **Calypso block editor**. This is the block editor loaded by Calypso on `https://wordpress.com/block-editor/post/<SITE_SLUG>`. It is actually the WP Admin block editor embed on an iframe. We also refer to this implementation as _Gutenframe_.
+ 
+## File architecture
 
-Server-side handlers of the different communication channels we establish with the client-side when Calypso loads the iframed block editor. See [`calypsoify-iframe.jsx`](https://github.com/Automattic/wp-calypso/blob/master/client/gutenberg/editor/calypsoify-iframe.jsx).
+- `/common`: Logic than runs on both environments (WP Admin and Calypso).
+- `/calypso`: Logic than runs only on the Calypso iframed block editor. 
+- `/wp-admin`: Logic than runs only on the WP Admin block editor.
```
Comment (kwight) [inline apps/wpcom-block-editor/README.md]: Might as well remove this, since it's not present, and not anticipated in the near future.

**Comment by kwight** [comment]
Awesome work ❤️ 

I did notice the minified scripts are missing from a manual build, and comparing the artifacts build versus the manual build shows a lot of differences, like maybe they're being built to different JS targets (`eval` vs. `module.exports`) – should they be aligned, and which would be preferable?

Screen Shot 2019-04-16 at 9 36 54 AM

> The documentation has also changed in order to make more obvious there are 2 environments we need to support. Note that I also grouped the previous scripts in a calypso/ folder to reflect that the utilities inside this folder will be run only in a iframed block editor context.

This is magnificent, thank you! 🎉

**Review by dmsnell — APPROVED** [review]
Looks like some good improvements here. I don't see any reason to block merging this.

Thanks @mmtr for updating this code. Those scripts are starting to make me long for our `Makefile` again… 😉

File: apps/wpcom-block-editor/package.json [hunk 2]
```diff
@@ -16,15 +16,19 @@
...
 	"scripts": {
-		"clean": "npx rimraf dist",
-		"prebuild": "npm run clean",
-		"bundle": "webpack --config='../../packages/calypso-build/webpack.config.js' --env.WP='true' iframe-bridge-server='./src/iframe-bridge-server.js' tinymce='./src/tinymce.js'",
-		"build": "npm run bundle -- --output-path=./dist"
+		"bundle": "webpack --config='../../packages/calypso-build/webpack.config.js' --env.WP='true'",
+		"build:common": "npm run bundle -- common='./src/common'",
+		"build:calypso-iframe-bridge-server": "npm run bundle -- calypso-iframe-bridge-server='./src/calypso/iframe-bridge-server.js'",
+		"build:calypso-tinymce": "npm run bundle -- calypso-tinymce='./src/calypso/tinymce.js'",
+		"build": "npm-run-all --parallel \"build:* -- {@}\" --"
```
Comment (simison) [inline apps/wpcom-block-editor/package.json]: If this gets too funky at some point, you could also create a simple Webpack config for this package that inherits and extends the one from Calypso-build, similarly to `client/notifications/`.

**Comment by mmtr** [comment]
> I did notice the minified scripts are missing from a manual build, and comparing the artifacts build versus the manual build shows a lot of differences (`eval` seems like not a great thing?) – should they be aligned, and which would be preferable?

Yeah, this is normal since the commands used in the manual build are different from the automatic CircleCi build.

The CircleCI job executes two commands:

1. Build for the unminified scripts.
```
SOURCEMAP='inline-cheap-source-map' npx lerna run build --scope='@automattic/wpcom-block-editor' -- -- --output-path=$CIRCLE_ARTIFACTS/wpcom-block-editor
```

`SOURCEMAP='inline-cheap-source-map'` is used for generating a more readable build file in order to ensure compliance with the plugin guidelines.

2. Build for the minified scripts.

```
NODE_ENV=production npx lerna run build --scope='@automattic/wpcom-block-editor' -- -- --output-path=$CIRCLE_ARTIFACTS/wpcom-block-editor --output-filename=[name].min.js
```

`NODE_ENV=production` is used for enabling the minification.

---

I was thinking in wrapping these 2 commands in a single npm script that replaces the current `build` script, so both builds (manual and automatic) will generate the same dist. But I didn't see any other script in Calypso that defines a `NODE_ENV` env var, which is the main reason why I didn't go for it, since this would prevent to skip the minification if we really don't want it. 

Probably in this case we want consistency rather than flexibility, so I'm going to commit a change grouping the 2 commands so `npx lerna run build --scope='@automattic/wpcom-block-editor' ` will generate both the unminified and minified bundles.

@sirreal @simison @ockham let me know if should take another approach here.

File: apps/wpcom-block-editor/package.json [hunk 3]
```diff
@@ -16,19 +16,25 @@
...
+		"@wordpress/editor": "9.2.2",
 		"@wordpress/hooks": "2.0.5",
+		"@wordpress/rich-text": "3.2.2",
 		"@wordpress/url": "2.3.3",
 		"jquery": "^1.12",
 		"lodash": "4.17.11",
 		"react": "16.8.6",
 		"tinymce": "4.8.5"
+	},
+	"devDependencies": {
```
Comment (sirreal) [inline apps/wpcom-block-editor/package.json]: A peculiarity of monorepo: `devDependencies` in non-root packages don't really mean anything. It's preferable to never include them, any devDependencies must be in the root package.

Comment (mmtr) [reply apps/wpcom-block-editor/package.json]: Wouldn't that cause a conflict with the `import/no-extraneous-dependencies` ESLint rule?

Screen Shot 2019-04-18 at 11 07 14

Comment (simison) [reply apps/wpcom-block-editor/package.json]: Might be. :-) It's not ideal — all I know it's causing trouble. See e.g.: https://github.com/Automattic/wp-calypso/pull/32383

cc @blowery

**Pull request merged**